reward/reward_distributor.go

```diff
@@ -38,113 +50,459 @@ type governanceHelper interface {
+func DistributeBlockReward(b BalanceAdder, rewards map[common.Address]*big.Int) {
+	for addr, amount := range rewards {
+		b.AddBalance(addr, amount)
 	}
 }
 
-// getTotalTxFee returns the total transaction gas fee of the block.
-func (rd *RewardDistributor) getTotalTxFee(header *types.Header, rewardConfig *rewardConfig) *big.Int {
-	totalGasUsed := big.NewInt(0).SetUint64(header.GasUsed)
-	if header.BaseFee != nil {
-		// magma hardfork
-		return totalGasUsed.Mul(totalGasUsed, header.BaseFee)
+func NewRewardConfig(header *types.Header, config *params.ChainConfig) (*rewardConfig, error) {
+	if config.Governance == nil {
+		return nil, errors.New("no config.Governance")
+	}
+	if config.Governance.Reward == nil {
+		return nil, errors.New("no config.Governance.Reward")
+	}
+
+	rules := config.Rules(header.Number)
+
+	cnRatio, kgfRatio, kirRatio, totalRatio, err := parseRewardRatio(config.Governance.Reward.Ratio)
+	if err != nil {
+		return nil, err
+	}
+
+	var cnProposerRatio, cnStakingRatio, cnTotalRatio int64
+	if rules.IsKore {
+		cnProposerRatio, cnStakingRatio, cnTotalRatio, err = parseRewardKip82Ratio(config.Governance.Reward.Kip82Ratio)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &rewardConfig{
+		// hardfork rules
+		rules: rules,
+
+		// values calculated from block header
+		totalFee: GetTotalTxFee(header, config),
+
+		// values from ChainConfig
+		mintingAmount: new(big.Int).Set(config.Governance.Reward.MintingAmount),
+		minimumStake:  new(big.Int).Set(config.Governance.Reward.MinimumStake),
+		deferredTxFee: config.Governance.Reward.DeferredTxFee,
+
+		// parsed ratio
+		cnRatio:    big.NewInt(cnRatio),
+		kgfRatio:   big.NewInt(kgfRatio),
+		kirRatio:   big.NewInt(kirRatio),
+		totalRatio: big.NewInt(totalRatio),
+
+		// parsed KIP82 ratio
+		cnProposerRatio: big.NewInt(cnProposerRatio),
+		cnStakingRatio:  big.NewInt(cnStakingRatio),
+		cnTotalRatio:    big.NewInt(cnTotalRatio),
+	}, nil
+}
+
+func GetTotalTxFee(header *types.Header, config *params.ChainConfig) *big.Int {
+	totalFee := new(big.Int).SetUint64(header.GasUsed)
+	if config.IsMagmaForkEnabled(header.Number) {
+		totalFee = totalFee.Mul(totalFee, header.BaseFee)
 	} else {
-		return totalGasUsed.Mul(totalGasUsed, rewardConfig.unitPrice)
+		totalFee = totalFee.Mul(totalFee, new(big.Int).SetUint64(config.UnitPrice))
 	}
+	return totalFee
 }
 
-func (rd *RewardDistributor) txFeeBurning(txFee *big.Int) *big.Int {
-	return txFee.Div(txFee, big.NewInt(2))
+// GetBlockReward returns the actual reward amounts paid in this block
+// Used in klay_getReward RPC API
+func GetBlockReward(header *types.Header, config *params.ChainConfig) (*RewardSpec, error) {
+	var spec *RewardSpec
+	var err error
+
+	if config.Istanbul == nil {
+		return nil, errors.New("no IstanbulConfig")
+	}
+
+	policy := config.Istanbul.ProposerPolicy
+	if policy == uint64(istanbul.RoundRobin) || policy == uint64(istanbul.Sticky) {
```

Comment:
in engine.go, the branch checks if the proposer policy is `WeightedRandom`. Could the checking routine be unified?

Reply:
@jiseongnoh The two ProposerPolicy-related branches are not the same. `GetBlockReward()` is not called from `engine.go:Finalize()`. `GetBlockReward()` is called from the new API to be added in #1666.

```
engine.go:Finalize() {
    if WeightedRandom {  // <-- the other one in engine.go
        CalcDeferredReward
    } else {
        CalcDeferredRewardSimple
    }
}

governance/api.go:GetRewards() {  // added in PR #1666
    GetBlockReward()
}
reward_distributor.go:GetBlockReward() {
    if RoundRobin || Sticky {  // <-- The one you asked
        CalcDeferredRewardSimple
    } else {
        CalcDeferredReward
    }
}
```

Reply:
I know they are called by different callers. However, I think they do the same thing (either calc or calcSimple). If they do the same thing, the branches can be in the same form, can't they? Is there any reason that the Finalize and getRewards should have different forms?

Reply:
@jiseongnoh Oh, I misunderstood. Thanks for the clarification, makes sense. @ian0371 Can we make a separate function like `shouldUseSimple(rewardConfig) bool`?

Reply:
let's address this in another PR